hail/python/hail/experimental/db.py

```diff
@@ -1,60 +1,284 @@
+    configuration.
+
+    Examples
+    --------
+    Create an annotation database connecting to the default Hail Annotation DB:
+
+    >>> db = hl.experimental.DB()
+    >>> mt = db.annotate_rows_db(mt, 'gnomad_lof_metrics') # doctest: +SKIP
+    """
+
+    _valid_key_properties = {'gene', 'unique'}
+
+    def __init__(self,
+                 url=None,
+                 json=None):
+        if json is not None and url is not None:
+            raise ValueError(f'Only specify one of the parameters url and json, '
+                             f'received: url={url} and json={json}')
+        if json is None:
+            if url is None:
+                url = ('https://www.googleapis.com/storage/v1/b/hail-common/o/annotationdb%2f' +
+                       str(hl.__pip_version__) +
+                       '%2fannotation_db.json?alt=media')
+            response = requests.get(url)
+            json = response.json()
+            assert isinstance(json, dict)
+        else:
+            if not isinstance(json, dict):
+                raise ValueError(f'expected a dict mapping dataset names to '
+                                 f'configurations, but found {json}')
+        self.__by_name = {k: Dataset.from_name_and_json(k, v)
+                          for k, v in json.items()}
+
+    def available_databases(self):
+        return self._by_name().keys()
+
     @staticmethod
-    def annotation_dataset_urls():
-        if DB._annotation_dataset_urls is None:
-            r = requests.get("https://www.googleapis.com/storage/v1/b/hail-common/o/annotationdb%2f1%2fannotation_db.json?alt=media")
-            j = r.json()
-            
-            DB._annotation_dataset_urls = {(x["name"], x.get("reference_genome")): x["path"] for x in j}
-            DB._geneDict_dataset_urls = { x["name"]: (x["gene_key"]) for x in j}
-        
-        return DB._annotation_dataset_urls, DB._geneDict_dataset_urls
-    
-    def annotate_rows_db(self,mt,*names):
+    def _row_lens(rel):
+        if isinstance(rel, hl.MatrixTable):
+            return RowLens(rel)
+        elif isinstance(rel, hl.Table):
+            return NoLens(rel)
+        else:
+            raise ValueError(
+                'annotation database can only annotate Hail MatrixTable or Table')
+
+    def dataset_by_name(self, name):
+        if name not in self.__by_name:
+            raise ValueError(
+                f'{name} not found in annotation database, you may list all '
+                f'known dataset names with available_databases()')
+        return self.__by_name[name]
+
+    def _annotate_gene_name(self, rel):
+        gene_field = Env.get_uid()
+        gencode = self.__by_name['gencode'].index_compatible_version(rel.key)
+        return gene_field, rel.annotate(**{gene_field: gencode.gene_name})
+
+    @typecheck_method(rel=oneof(table_type, matrix_table_type), names=str)
+    def annotate_rows_db(self, rel, *names):
```

Comment:
I don't totally love `db.annotate_rows_db`, but don't have anything better to suggest. I would prefer `db.annotate` in a world where `annotate` isn't the "assign fields" method on Table/MT.

Reply:
I prefer `annotate`. I am hesitant to change the method name given that people are using this already. When I promote it out of `experimental`, I hoped to change the name to `annotate` (and provide a deprecated shim in `experimental/db.py`).